DISTINCT in order optimization

devcrafter opened the pull request:
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Distinct optimization for sorted columns. Use specialized distinct transformation in case input stream is sorted by column(s) in distinct. Optimization can be applied to pre-distinct, final distinct, or both. Initial implementation by @dimarub2000 

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

devcrafter committed:
DISTINCT in order optimization

+ optimization for DISTINCT containing primary key columns

alexey-milovidov commented:
Let's also reference the original author and the original pull request in the changelog (we have this practice if the original author has done more than 0% of the work).

devcrafter commented:
@nickitat please could you review it?

nickitat commented:
notes regarding the approach:
* distinct usually happens in two stages: multi-threaded pre-distinct and single-threaded final distinct. and it doesn't make much practical sense to optimize only pre-distinct, because, in case of high-cardinality data, execution time will be dominated by final distinct. so, IMO, if optimization is applicable on the final stage - we should apply it.
* on the road to DISTINCT transform, data may be already reordered. in case of pre-distinct it may happen in query like `select distinct n from t group by n`, in case of final distinct - in query like `select distinct n from t order by sipHash64(n)`. probably we should look into `input_streams.front().sort_description` in DistinctStep when adding transforms into pipeline. 
* to read data initially in-order we need to choose the special type of reading for that as we do [for aggregation](https://github.com/ClickHouse/ClickHouse/blob/master/src/Processors/QueryPlan/ReadFromMergeTree.cpp#L978-L980). and propagate info about sorting order through `sort_description` [here](https://github.com/ClickHouse/ClickHouse/blob/master/src/Processors/QueryPlan/ReadFromMergeTree.cpp#L82-L86).

devcrafter committed:
DISTINCT in order optimization

+ use SortDescription from input data stream in DistinctStep to decide if the optimization is applicable

devcrafter commented:
@Mergifyio update

mergify[bot] committed:
Merge branch 'master' into dictinct_in_order_optimization

devcrafter committed:
DISTINCT in order optimization

+ try use the optimization for final distinct in case of sorted stream
  (sorting inside and among chunks)
+ sorting description contains only columns from sorting key which are in
  header as well

devcrafter commented:
@Mergifyio update

mergify[bot] committed:
Merge branch 'master' into dictinct_in_order_optimization

devcrafter committed:
Add plan tests for final distinct

devcrafter committed:
Merge remote-tracking branch 'origin/master' into dictinct_in_order_optimization

devcrafter committed:
Fix: flaky test

devcrafter committed:
Fixes
+ flaky test with explain pipeline
+ consider sort direction from read order info in sort description
  (ReadFromMergeTree step)

devcrafter commented:
@Mergifyio update

mergify[bot] committed:
Merge branch 'master' into dictinct_in_order_optimization

devcrafter committed:
Remove unnecessary include

devcrafter committed:
Fixes:
+ test warning
+ proper capacity for column positions array in DistinctTransform

devcrafter committed:
Merge remote-tracking branch 'origin/master' into dictinct_in_order_optimization

devcrafter committed:
Adapt distinct for sorted chunks to handle sorted stream, so we can use
it for final distinct as well

devcrafter commented:
@nickitat Please take a look. Thanks

devcrafter commented on tests/queries/0_stateless/02317_distinct_in_order_optimization_explain.sh:
```diff
@@ -0,0 +1,43 @@
+
+DISABLE_OPTIMIZATION="set optimize_distinct_in_order=0"
+ENABLE_OPTIMIZATION="set optimize_distinct_in_order=1"
+GREP_OPTIMIZATIONS="grep 'DistinctSortedChunkTransform\|DistinctSortedTransform'"
```
`DistinctSortedTransform` need to be deleted here.

devcrafter commented on tests/queries/0_stateless/02317_distinct_in_order_optimization_explain.reference:
```diff
@@ -0,0 +1,19 @@
+-- distinct with non-primary key prefix -> no optimizations
+No optimizations
+-- distinct with non-primary key prefix and order by on column in distinct -> final distinct optimization only
+DistinctSortedChunkTransform
```
I'd prefer to have a different name for final distinct, i.e. `DistinctSortedStreamTransform`. But the implementation is the same, so I'm not sure if it is worth the hassle to have a flag (i.e.`final_distinct`), and provide the different name if it's true. @KochetovNicolai any thoughts?

devcrafter committed:
Some polishing

devcrafter committed:
Fix: input chunk can have empty columns (no rows)

devcrafter commented:
@mergifyio update

mergify[bot] committed:
Merge branch 'master' into dictinct_in_order_optimization

devcrafter committed:
Mark condition for empty chunk as unlikely

nickitat commented on src/Processors/Transforms/DistinctSortedChunkTransform.h:
```diff
@@ -0,0 +1,63 @@
+/// DISTINCT optimization for sorted chunks
+/// It also checks if current chunks is continuation of previous one,
+/// i.e. sorting prefix value of last row in previous chunk is the same as of first row in current one,
+/// so it can correctly process sorted stream as well
```
it worth to mention here the logic depending on `other_columns.empty()`. if true, we will go without hash tables at all. if false - HT will be created for each value of sorting key prefix which means lower memory consumption.

nickitat reviewed (commented):
we also need perf tests. with low or high cardinality data x only pre-distinct or both x with or without ordinary distinct under the hood

devcrafter committed:
Fixes for some review comments

devcrafter committed:
Perf tests with high cardinality

devcrafter commented:
@Mergifyio update

mergify[bot] committed:
Merge branch 'master' into dictinct_in_order_optimization

devcrafter committed:
Adapt range search algorithm to high cardinality case

+ range search done in steps of some number of rows.
  Controled by new
  setting `distinct_in_order_range_search_step`. By default 0, i.e.
  whole chunk
+ before start binary search, linear probing is done on each step (32
  rows currently)

devcrafter commented:
@Mergifyio update

mergify[bot] committed:
Merge branch 'master' into dictinct_in_order_optimization

devcrafter committed:
+ use DistinctSorted for final distinct step
+ fix performance tests

nickitat commented on src/Processors/Transforms/DistinctSortedChunkTransform.cpp:
```diff
@@ -0,0 +1,232 @@
+    ///     otherwise: apply ordinary distinct
+    /// (3) repeat until chunk is processed
+    IColumn::Filter filter(chunk_rows);
+    auto [range_begin, output_rows] = continueWithPrevRange(chunk_rows, filter); /// try to process chuck as continuation of previous one
```
I suggest to rename `range_begin` to `next_range_begin`

nickitat commented:
perf tests results look really good.
I think speed up might be even bigger if we apply read in order. but it makes sense to do it out of scope of this pr.

devcrafter committed:
Addressing review comments

devcrafter commented:
@Mergifyio update

mergify[bot] committed:
Merge branch 'master' into dictinct_in_order_optimization

devcrafter commented on src/Processors/Transforms/DistinctSortedChunkTransform.cpp:
```diff
@@ -123,10 +122,23 @@ bool DistinctSortedChunkTransform::isCurrentKey(const size_t row_pos) const
+            return pos;
+    }
+
+    size_t low = linear_probe_end;
```
I'd check it in a follow-up PR
